lib/fake_ifdh.py

```diff
@@ -94,21 +102,19 @@ def checkToken(tokenfile: str) -> bool:
     """check if token is (almost) expired"""
     if not os.path.exists(tokenfile):
         return False
-    exp_time = None
-    cmd = f"decode_token.sh -e exp {tokenfile} 2>/dev/null"
-    with os.popen(cmd, "r") as f:
-        exp_time = f.read()
     try:
-        add_event(f"expiration: {exp_time}")
-        return int(exp_time) - time.time() > 60
-    except ValueError as e:
-        print(
-            "decode_token.sh could not successfully extract the "
-            f"expiration time from token file {tokenfile}. Please open "
-            "a ticket to Distributed Computing Support if you need further "
-            "assistance."
-        )
-        raise
+        exp_time = None
+        with open(tokenfile) as f:
+            token_encoded = f.read(4096).strip()
+            token = dict(
+                scitokens.SciToken.deserialize(token_encoded, insecure=True).claims()
+            )
+            exp_time = token["exp"]
+            add_event(f"expiration: {exp_time}")
+            return int(exp_time) - time.time() > 60
+    except Exception:
```

Comment:
What are your thoughts about having two kinds of `Exception` here, one clause for `scitokens.SciTokensException` and the other for general `Exception`?  That way we could even test both perhaps.

Reply:
@marcmengel  and @retzkek:  So this reraise solution I _don't_ particularly like.  It strikes me as a bit of a code smell and we won't remember six months from now why we did it and why `checkToken` has this random `reraise` bool.

Instead, I do like the original idea of Kevin's that the caller (which is just the two times in `getToken`) knows what to do.  I think it'd keep the intended return states of `checkToken` much clearer.  If we didn't want to change much, we could only change the `getToken` code so it checks for `True` or `False` the first go-around (like it currently does), and the second time, it will do a try-except or something like that.

If we want to get a bit more adventurous, but IMO more clear in the long run, and since the impact is pretty localized, we could even change the signature of `checkToken` so it just returns None, and raises an `Exception` in any error case or a failure case, which `getToken` would know how to handle.  

For example, `checkToken` currently only returns `False` if the tokenfile doesn't exist, or fails the time to expiration check.  Both of these could be raised `Exceptions` (an `OSError` for the first, and maybe a custom `TokenExpirationError` for the second), which `getToken` would know how to handle.  Your current restructuring of the code, Marc, actually allows this to be implemented pretty cleanly.  Then, if `checkToken` raises any other exception, we'd reraise that (or reraise ANY `Exception` in the second case).

I think that'd look something like this (pseudocode):

```python

def checkToken(tokenfile: str) -> None:
    exp_time = None
    try:
        with open(tokenfile) as f:
            do_stuff()
            assert int(exp_time) - time.time() > 60
    except OSError:
        # File not found
        raise
    except AssertionError as e:
        raise TokenExpirationError("Token expires less than 60 seconds from now, at {string_for_time}")
    # Other future exception clauses?
    except Exception as e:
        raise

def getToken(role: str = DEFAULT_ROLE, verbose: int = 0) -> str:
    do_stuff()
    # Change this call since we'd return None
    try:
        checkToken(tokenfile)
    except (OSError, TokenExpirationError) as e:
        # maybe print the exception so in verbose mode we know what's going on?
        if verbose > 0:
            print(e). # Something better than this
        do_the_htgettoken_stuff_including_raising_PermissionError_if_it_fails()
        checkToken(tokenfile)  # This will automatically reraise if something fails this time
        return tokenfile
    except Exception as e:
        raise e
    else:
        return tokenfile
```

This is a long bit of pseudocode, and yes, @retzkek , a little Go-ish.  I'm happy to type it up for real and put it in here if you guys think it's a good idea.  I think it'll make it very clear that `checkToken` either returns nothing if the check works, or some kind of Exception if it fails, and then getToken handles any reraising or not, which satisfies Kevin's initial point (one I agree with).

What are your thoughts?

Reply:
So basically now checkToken always raises exceptions, and in the place we want to keep going if it fails, we do a generic except: that just says we need to fetch a new token.    THen having tried to fetch one, if checkToken raises anything we let that kill us.